File: src/libawkward/type/UnknownType.cpp
// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/master/LICENSE

#define FILENAME(line) FILENAME_FOR_EXCEPTIONS("src/libawkward/type/UnknownType.cpp", line)

#include <string>
#include <sstream>

#include "awkward/array/EmptyArray.h"
#include "awkward/type/UnknownType.h"

namespace awkward {
  UnknownType::UnknownType(const util::Parameters& parameters,
                           const std::string& typestr)
      : Type(parameters, typestr) { }

  std::string
  UnknownType::tostring_part(const std::string& indent,
                             const std::string& pre,
                             const std::string& post) const {
    std::string typestr;
    if (get_typestr(typestr)) {
      return wrap_categorical(typestr);
    }

    std::stringstream out;
    if (parameters_empty()) {
      out << indent << pre << "unknown" << post;
    }
    else {
      out << indent << pre << "unknown[" << string_parameters() << "]" << post;
    }
    return wrap_categorical(out.str());
  }

  const TypePtr
  UnknownType::shallow_copy() const {
    return std::make_shared<UnknownType>(parameters_, typestr_);
  }

  bool
  UnknownType::equal(const TypePtr& other, bool check_parameters) const {
    if (UnknownType* t = dynamic_cast<UnknownType*>(other.get())) {
      if (check_parameters  &&  !parameters_equal(other.get()->parameters())) {
        return false;
      }
      return true;
    }
    else {
      return false;
    }
  }

  int64_t
  UnknownType::numfields() const {
    return -1;
  }

  int64_t
  UnknownType::fieldindex(const std::string& key) const {
    throw std::invalid_argument(
      std::string("type contains no Records") + FILENAME(__LINE__));
  }

  const std::string
  UnknownType::key(int64_t fieldindex) const {
    throw std::invalid_argument(
      std::string("type contains no Records") + FILENAME(__LINE__));
  }

  bool
  UnknownType::haskey(const std::string& key) const {
    throw std::invalid_argument(
      std::string("type contains no Records") + FILENAME(__LINE__));
  }

  const std::vector<std::string>
  UnknownType::keys() const {
    throw std::invalid_argument(
      std::string("type contains no Records") + FILENAME(__LINE__));
  }

  const ContentPtr
  UnknownType::empty() const {
    return std::make_shared<EmptyArray>(Identities::none(), parameters_);
  }
}

File: src/awkward/_v2/types/unknowntype.py
# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE

from __future__ import absolute_import

import awkward as ak
from awkward._v2.types.type import Type
from awkward._v2.forms.form import _parameters_equal


class UnknownType(Type):
    def __init__(self, parameters=None, typestr=None):
        if parameters is not None and not isinstance(parameters, dict):
            raise TypeError(
                "{0} 'parameters' must be of type dict or None, not {1}".format(
                    type(self).__name__, repr(parameters)
                )
            )
        if typestr is not None and not ak._util.isstr(typestr):
            raise TypeError(
                "{0} 'typestr' must be of type string or None, not {1}".format(
                    type(self).__name__, repr(typestr)
                )
            )
        self._parameters = parameters
        self._typestr = typestr

    def __str__(self):
        if self._typestr is not None:
            out = self._typestr

        else:
            params = self._str_parameters()
            if params is None:
                out = "unknown"
            else:
                out = "unknown[" + params + "]"

        return self._str_categorical_begin() + out + self._str_categorical_end()

    def __repr__(self):
        args = self._repr_args()
        return "{0}({1})".format(type(self).__name__, ", ".join(args))

    def __eq__(self, other):
        if isinstance(other, UnknownType):
            return self._typestr == other._typestr and _parameters_equal(
                self._parameters, other._parameters, only_array_record=True
            )
        else:
            return False
